### ISSUE TITLE
Implement backup threads for domains

### ISSUE THREAD

**Issue by kayceesrk** [ocaml-multicore/ocaml-multicore#90]
**tl;dr** Instantiate multiple system threads to service a single domain

In the current design, every domain is associated with a single system thread that services the domain. Whenever a domain blocks on a external call, the lock on the domain is released so that another domain may temporarily take over the domain to:

- perform major GC on behalf of the blocked thread
- service promotion requests

The fact that a single system thread might be responsible for managing multiple domains at the same time is tricky and leads to subtle bugs. 

The proposal is to create 2 system threads for each domain, one which is the primary mutator thread and runs OCaml code and another is a backup thread which only steps in when the primary thread is blocked on an external call. By default, the backup thread sleeps on a condition variable protected by the domain lock and is woken up to service GC or promotion requests.

**Comment by kayceesrk** [comment]
Should be done after #146.

**Comment by kayceesrk** [comment]
See #356

**Comment by ctk21** [comment]
This issue is stale and backup threads got merged in #356.
